### PR TITLE
Added usability to CoproductInstances

### PR DESCRIPTION
**Pull request by Kalin-Rudnicki**

### CODE REVIEW

File: modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala [hunk 1]
```diff
@@ -305,8 +305,8 @@ object ErasedProductInstances:
       case 1 => ErasedProductInstances1[K, FT](mirror, summonOne[E])
       case _ => ErasedProductInstancesN[K, FT](mirror, summonAsArray[E])
 
-final class ErasedCoproductInstances[K, FT](mirror: Mirror.Sum, is0: => Array[Any]) extends ErasedInstances[K, FT]:
-  private lazy val is: Array[Any] = is0
+final class ErasedCoproductInstances[K, FT](val mirror: Mirror.Sum, is0: => Array[Any]) extends ErasedInstances[K, FT]:
+  lazy val is: Array[Any] = is0
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/internals/erased.scala:309]: I would prefer if we don't expose these

File: modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala [hunk 2]
```diff
@@ -150,10 +150,18 @@ object K0:
       inst.asInstanceOf
 
   extension [F[_], T](inst: CoproductInstances[F, T])
+    inline def instanceIndex(t: T): Int =
+      inst.mirror.ordinal(t.asInstanceOf)
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala:154]: The signature here is wrong `T` is the type class, not the enum so we can't get an ordinal from the mirror. Also what is we just exposed `mirror.ordinal` in the erased instances?

File: modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala [hunk 3]
```diff
@@ -150,10 +150,18 @@ object K0:
...
+    inline def instanceFromIndex(idx: Int): F[T] =
+      inst.is(idx).asInstanceOf[F[T]]
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala:156]: This is just `inject(identity)`

File: modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala [hunk 4]
```diff
@@ -150,10 +150,18 @@ object K0:
...
+    inline def instanceFor(t: T): F[T] =
+      instanceFromIndex(instanceIndex(t))
     inline def mapK[G[_]](f: [t <: T] => F[t] => G[t]): CoproductInstances[G, T] =
       inst.erasedMapK(f.asInstanceOf).asInstanceOf
     inline def inject[R](p: Int)(f: [t <: T] => F[t] => R): R =
       inst.erasedInject(p)(f.asInstanceOf).asInstanceOf
+    inline def injectFor[R](t: T)(f: [t] => (F[t], t) => R): R =
+      inst.inject[R](instanceIndex(t))([t] => (i: F[t]) => f[t](i, t.asInstanceOf[t]))
```
Comment (joroKr21) [inline modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala:164]: This is just `fold`

**Comment by Katrix** [comment]
These functions widens what I see as a soundness hole in shapeless 3. Shapeless 3 uses ints to index into it's types, which without these functions is somewhat okay as getting these ints is not straightforward. These functions do make it straight forward however. 

The problem is that the type `F[T]` obtained through indexing into the types is a lie, and is secretly `F[_ <: T]`. I don't know what the right answer for shapeless is. My own library, perspective, uses various tricks to ensure soundness, (existentials<sup>[1]</sup> and path dependent types<sup>[2] [3]</sup>). These I think work, but reduce usability and friendliness.

For `instanceFor`, `inline def instanceFor[A <: T](t: A): F[A]` might seem like a better bet. In in some ways it is, but it can still be used in unsound ways. Either the argument needs to exclude values of type `T`, or the return type needs to hide something about what it is for.

[1]: https://github.com/Katrix/perspective/blob/main/dotty/derivation/src/main/scala/perspective/derivation/hkdGeneric.scala#L331
[2]: https://github.com/Katrix/perspective/blob/main/dotty/derivation/src/main/scala/perspective/derivation/hkdGeneric.scala#L338
[3]: https://github.com/Katrix/perspective/blob/main/dotty/derivation/src/main/scala/perspective/derivation/hkdGeneric.scala#L389-L402

**Comment by joroKr21** [comment]
All very good points 👍 

> The problem is that the type F[T] obtained through indexing into the types is a lie, and is secretly F[_ <: T]. I don't know what the right answer for shapeless is.

It sounds like the answer would be to return `F[_ <: T]` (if we add this function at all)

**Comment by Katrix** [comment]
If Scala 3 doesn't throw a fuzz about unreducible application of higher kinded types or whatever, then yes, that's the simplest approach.

**Pull request closed without merging**